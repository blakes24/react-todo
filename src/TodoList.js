import { useState } from 'react';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import { v4 as uuidv4 } from 'uuid';
import './TodoList.css';

const TodoList = () => {
	const [ todos, setTodos ] = useState([]);
	const [ editForm, setEditForm ] = useState({ value: '', id: '' });

	const remove = (id) => {
		const filteredTodos = todos.filter((todo) => todo.id != id);
		setTodos(filteredTodos);
	};
	const addTodo = (formData) => {
		const newTodo = { ...formData, id: uuidv4() };
		setTodos((todos) => [ ...todos, newTodo ]);
	};
	const edit = (formData, id) => {
		const filteredTodos = todos.filter((todo) => todo.id != id);
		const updatedTodo = { ...formData, id: id };
		setTodos(() => [ ...filteredTodos, updatedTodo ]);
		setEditForm(() => ({ value: '', id: '' }));
	};
	const showForm = (id, todo) => {
		setEditForm({ value: todo, id: id });
	};

	return (
		<div className="TodoList">
			<h1>Todo List</h1>
			<NewTodoForm submitFunc={addTodo} placeholder="Add a todo" label="New todo:" value="" btn="Add!" />
			{editForm.value.length > 0 && (
				<NewTodoForm
					submitFunc={edit}
					label="Edit todo:"
					value={editForm.value}
					id={editForm.id}
					btn="Submit"
				/>
			)}
			{todos.map(({ id, todo }) => <Todo id={id} todo={todo} remove={remove} showForm={showForm} key={id} />)}
		</div>
	);
};

export default TodoList;
