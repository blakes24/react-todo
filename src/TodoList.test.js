import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

it('renders without crashing', function() {
	render(<TodoList />);
});

it('matches snapshot', function() {
	const { asFragment } = render(<TodoList />);
	expect(asFragment()).toMatchSnapshot();
});

it('can add a todo', function() {
	const { getByLabelText, queryByText } = render(<TodoList />);

	// no todos yet
	expect(queryByText('new todo')).not.toBeInTheDocument();

	const todoInput = getByLabelText('New todo:');
	const submitBtn = queryByText('Add!');

	// fill out the form
	fireEvent.change(todoInput, { target: { value: 'new todo' } });
	fireEvent.click(submitBtn);

	// todo exists
	expect(queryByText('new todo')).toBeInTheDocument();
});

it('can remove a todo', function() {
	const { getByLabelText, queryByText } = render(<TodoList />);

	const todoInput = getByLabelText('New todo:');
	const submitBtn = queryByText('Add!');

	// fill out the form
	fireEvent.change(todoInput, { target: { value: 'new todo' } });
	fireEvent.click(submitBtn);

	// todo exists
	expect(queryByText('new todo')).toBeInTheDocument();

	// click delete button
	fireEvent.click(queryByText('X'));

	// no todo in doc
	expect(queryByText('new todo')).not.toBeInTheDocument();
});

it('can edit a todo', function() {
	const { getByLabelText, queryByText, queryByDisplayValue } = render(<TodoList />);

	const todoInput = getByLabelText('New todo:');
	const submitBtn = queryByText('Add!');

	// add a todo
	fireEvent.change(todoInput, { target: { value: 'new todo' } });
	fireEvent.click(submitBtn);
	expect(queryByText('new todo')).toBeInTheDocument();

	// click edit button
	fireEvent.click(queryByText('Edit'));

	// edit todo
	const editInput = queryByDisplayValue('new todo');
	fireEvent.change(editInput, { target: { value: 'edited todo' } });
	fireEvent.click(queryByText('Submit'));

	expect(queryByText('edited todo')).toBeInTheDocument();
	expect(queryByText('new todo')).not.toBeInTheDocument();
});

it('clears form fields on submit', function() {
	const { getByLabelText, queryByText, queryByDisplayValue } = render(<TodoList />);

	const todoInput = getByLabelText('New todo:');
	const submitBtn = queryByText('Add!');

	// fill out the form
	fireEvent.change(todoInput, { target: { value: 'new todo' } });
	expect(queryByDisplayValue('new todo')).toBeInTheDocument();

	// submit form
	fireEvent.click(submitBtn);

	// form field is empty
	expect(queryByDisplayValue('new todo')).not.toBeInTheDocument();
});
