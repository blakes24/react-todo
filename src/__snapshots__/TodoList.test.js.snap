// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <h1>
      Todo List
    </h1>
    <form>
      <label
        for="todo"
      >
        New todo: 
      </label>
      <input
        id="todo"
        name="todo"
        placeholder="Add a todo"
        type="text"
        value=""
      />
      <button>
        Add!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
